src/lib/upgradePackageDefinitions.ts

```diff
@@ -49,18 +127,47 @@ export async function upgradePackageDefinitions(
 
   if (options.peer && Object.keys(filteredLatestDependencies).length > 0) {
     const upgradedPeerDependencies = await getPeerDependenciesFromRegistry(filteredLatestDependencies, options)
-    const peerDependencies = { ...options.peerDependencies, ...upgradedPeerDependencies }
-    if (!dequal(options.peerDependencies, peerDependencies)) {
-      const [newUpgradedDependencies, newLatestVersions, newPeerDependencies] = await upgradePackageDefinitions(
-        { ...currentDependencies, ...filteredUpgradedDependencies },
-        { ...options, peerDependencies, loglevel: 'silent' },
-      )
-      return [
-        { ...filteredUpgradedDependencies, ...newUpgradedDependencies },
-        { ...latestVersionResults, ...newLatestVersions },
-        newPeerDependencies,
-      ]
+
+    let rerunResult: RerunUpgradeResult
+    let checkPeerViolationResult: CheckIfInPeerViolationResult = {
+      violated: false,
+      filteredUpgradedDependencies,
+      upgradedPeerDependencies,
     }
+    let runIndex = 0
+    do {
+      rerunResult = await rerunUpgradeIfChangedPeers(
+        currentDependencies,
+        checkPeerViolationResult.filteredUpgradedDependencies,
+        checkPeerViolationResult.upgradedPeerDependencies,
+        latestVersionResults,
+        options,
+      )
+      if (!rerunResult) {
+        if (runIndex > 0) {
+          // We can't find anything to do, will not upgrade anything
+          return [{}, latestVersionResults, options.peerDependencies]
+        }
+        checkPeerViolationResult = checkIfInPeerViolation(
+          currentDependencies,
+          filteredUpgradedDependencies,
+          upgradedPeerDependencies,
+        )
+        if (!checkPeerViolationResult.violated) {
+          // No issues were found, return
+          return [filteredUpgradedDependencies, latestVersionResults, options.peerDependencies]
+        }
+      } else {
+        checkPeerViolationResult = checkIfInPeerViolation(currentDependencies, rerunResult[0], rerunResult[2]!)
+        if (!checkPeerViolationResult.violated) {
+          // We found a stable solution
+          return rerunResult
+        }
+      }
+      if (runIndex++ > 5) {
+        throw new Error(`Stuck in a while loop. Please report an issue`)
+      }
+    } while (true)
```

Comment:
This would make more sense as `while (checkPeerViolationResult.violated)` since the dependencies will keep getting checked until there are no more violations.

Reply:
Done